Modbus patch, to allow communication with "slow" equipment using tcp

janiversen opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Some modbus-tcp servers (like e.g. Huawei sun2000) need time between the connect() and the first send(), in order to prepare themself.

A typical error pattern is that connect() works perfectly, but when requesting data there are no response (apart from the TCP/IP ACK message). Once in this mode, the server cannot recover and later messages are also not responded to.

The modbus configuration (only for modbus-tcp) have been extended with a new optional parameter "delay" with default being 0. When adding e.g. delay: 2, time.sleep(2) will be called directly after the connect() returns. If delay is not present in the configuration it has a value of 0, and time.sleep() are not called.

Another PR have been filed to update the documentation.

There should not be any breaking changes, basically it is "just" a change from
sync to async. However the library version is bumped from 1.5.2 to 2.3.0, and
that might solve/cause problems. 

This started as a 2-3 line change, but now something like 30% of the code is touched,
due to the request of making it async.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml
modbus:
    name: sun2000
    type: tcp
    host: 192.168.0.3
    port: 502
    delay: 2

sensor:
  platform: modbus
  scan_interval: 10
  registers:
    - name: powerMeter
      hub: sun2000
      unit_of_measurement: W
      slave: 0
      register: 37113
      count: 2
```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
The PR is done as 2 commits.

First commit adds a fixed time.sleep(2), which is the solution (apart from the 2)
Second commit adds the additional items needed to make time.sleep(_delay) configurable

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
[https://github.com/home-assistant/home-assistant.io/pull/12305](url)

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x ] The code change is tested and works locally.
- [ x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ x] There is no commented out code in this PR.
- [ x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

homeassistant commented:
Hi @janiversen, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#32557).

Once you do that we will be able to review and accept this pull request.

Thanks!

janiversen commented:
It would be nice if someone could take a look at this small patch, so it can (hopefully) get merged, and used.

Emilv2 commented:
I think the devs have said that time.sleep is never allowed, but I could be wrong.

I'm interested in your usecase, since I also had some trouble getting modbus to read out the Huawei sun2000, and I seem to have different results.

If I understand your code correctly, the connect (and delay) is only used once during HA startup, after which there are no problems anymore?
My device frequently gives errors, and the solution I found is to try to read out three times, it always works on the third try. Delay between the three reads does not matter (unless the timeout is small, then more might be needed). Immediately following reads work from the first try, but even after several seconds three tries are needed again. Not really a very elegant solution of course.

Waiting after connect seems to work, but I still need to repeat that frequently (every ~10 seconds) or the problem reoccurs.

janiversen commented:
I agree that using time.sleep in the running code is something that should be avoided and are normally not really needed since the worker thread will call every second. However in setup once you return (unless I am mistaken, which very well might be the case) setup is done, so because of that a wait is needed in the function.

If you poll the sun2000 too fast it will fail. Do you use the scan interval as pr my example? It runs for me andI might get a read error 2-3 times a day but not more.

Can you please provide the modbus part of your config then I can try it on my installation and see what can be done

Emilv2 commented:
I don't use the build-in modbus component but I build my own: https://github.com/Emilv2/huawei_solar
The library doing the actual communication is here: https://gitlab.com/Emilv2/huawei-solar

I quickly tried your solution using pymodbus manually on an interactive python, and your solution seems to work. Close the connection, connect the client, and then wait two seconds to read the registers, but I need to repeat this every time I start to read registers. Polling it very fast doesn't to fail for me, it's only the first read in a sequence that fails.

I would say that if I could get async pymodbus to work that would be perfect, since `asyncio.sleep()` is non-blocking and allowed. But for some reason I can't get it to work: https://community.home-assistant.io/t/async-pymodbus/177601

janiversen commented:
Why do you need to connect for every poll cycle? that creates a bit of stress in the sun2000 and seems to be a lot of overhead.

How does the first one fail? no response at all, only a tcp ack or a error message? I have a non home assistant python3 script that works without problems. The Huawei developers have confirmed that a delay after connect is needed.

I am actually testing a new integration with async, but async.sleep() does not solve the delay after connect since it is in setup(). I read multiple registers in one read and then have a short async.sleep(). So far I have not had problems with the core of home assistant. I made the setup async as well.

janiversen commented:
Just had a fast look at your code, I miss a thread lock is that on purpose?

I will try your code this weekend, it looks a lot like my code, with 2 big differences. Based on the configuration I group the registers and read then in blocks and I keep a local value to further limit the number of actual reads.

Emilv2 commented:
Because otherwise it fails just like if I never did the sleep. I think it was my HA instance reading out at the same time that made it do that, with it disabled it works as you described. multiple connections at the same time don't seem to be supported.

Without the sleep it always fails in the same way like this:
```
>>> result = client.read_holding_registers(30000,15)
ModbusIOException(InvalidMessageReceivedException('Incomplete message received, expected at 
least 8 bytes (0 received)'), 3)
>>> result = client.read_holding_registers(30000,15)
ModbusIOException('No Response received from the remote unit/Unable to decode response', 3)
>>> result = client.read_holding_registers(30000,15)
>>> result.encode()
b'\x1eSUN2000L-3KTL\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
```

I'm fairly new to asyncio, where would I need a thread lock?

Which local variables do you keep that I don't? The variables that do not change such as name and serial numbers I poll only once and the ones that change I poll. I thought about reading the registers in blocks, but that seems to make it so much more complicated.

janiversen commented:
the thread lock is for both sync and async, it should be around each pymodbus call. Sun2000 do not support multiple connections, and if you open/close-open/close without a larger delay you will rum into problems.

asyncio is a bit more complicated, just as reading multiple registers with on read call, but it pays to do it.

I keep local variables for all registers, because from a automation pow, only current production and the import/value matters. I actually build a third variable “current comsumption” = production - power meter (import/export).

Emilv2 commented:
I tried on three different devices, and all keep the same problem:
```
import pymodbus
import time
client = pymodbus.client.sync.ModbusTcpClient('192.168.0.135', '502')
client.connect()
time.sleep(3)
# works
result = client.read_holding_registers(30000,15)
time.sleep(180) # not sure about the exact value needed
# fails again
result = client.read_holding_registers(30000,15)
```

My code does not get stuck on any async part, it gets stuck on the creation of the modbus client. I can try it, but I don't see how a lock can help there?

janiversen commented:
the lock is not the connect() but for the reads, however since you reconnect every time, you are very likely to run into problems with multiple connects simultaneously.

As I wrote the lock is not to solve your async problem, since a lock is also needed for the sync calls. Your async problems are different, and I will try to compare my async code to see why yours might fail.

Emilv2 commented:
I don't think adding a lock would make sense in my sync code? All the updates are done sequentially in the same function.

While the current production is the most useful for automation, the other values can be of use too for monitoring your setup so I poll them too.

bdraco commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -206,6 +211,8 @@ def connect(self):
         """Connect client."""
         with self._lock:
             self._client.connect()
+            if self._delay > 0:
```
Please adjust the timeout instead of sleeping

see ` pymodbus/client/sync.py `
```
# --------------------------------------------------------------------------- #
# Modbus TCP Client Transport Implementation
# --------------------------------------------------------------------------- #
class ModbusTcpClient(BaseModbusClient):
    """ Implementation of a modbus tcp client
    """

    def __init__(self, host='127.0.0.1', port=Defaults.Port,
        framer=ModbusSocketFramer, **kwargs):
        """ Initialize a client instance

        :param host: The host to connect to (default 127.0.0.1)
        :param port: The modbus port to connect to (default 502)
        :param source_address: The source address tuple to bind to (default ('', 0))
        :param timeout: The timeout to use for this socket (default Defaults.Timeout)
        :param framer: The modbus framer to use (default ModbusSocketFramer)

        .. note:: The host argument will accept ipv4 and ipv6 hosts
        """
        self.host = host
        self.port = port
        self.source_address = kwargs.get('source_address', ('', 0))
        self.socket = None
        self.timeout = kwargs.get('timeout',  Defaults.Timeout)
        BaseModbusClient.__init__(self, framer(ClientDecoder(), self), **kwargs)
```

bdraco replied:
If the timeout doesn't fix it you might need to do a select with a timeout on the socket to verify its ready

See _recv in ModbusTcpClient
```
            ready = select.select([self.socket], [], [], end - time_)
```

Emilv2 replied:
@bgianfo Adjusting timeout doesn't work for this device (Huawei SUN2000). You need to wait between connecting and sending (or try three times in a row).

janiversen replied:
I would love to avoid the sleep, but the timeout parameter works on the socket before the connect, and as @Emilv2 noted the delay is after a succesfull connect() and the first send().

If there is a function in core like wait(secs) please tell me and I will implement that.

bdraco replied:
This can't be async with a sync client.  This would need to be converted to async

```
from pymodbus.client.sync import ModbusSerialClient, ModbusTcpClient, ModbusUdpClient
```

pymodbus has an async client.
https://pymodbus.readthedocs.io/en/latest/source/library/pymodbus.client.html

Emilv2 commented:
Shouldn't `async.sleep()` work as well when you're using the lock during the sleep? The component should be first changed to use async.

janiversen commented:
I do not think so, it will make the connection sleep, but the thread will continue, meaning that setup() will terminate and the update process will start.

I did a short test to verify this, and I can change it to:
await asyncio.sleep(secs)
but that seems to the same as time.sleep(secs).
Remark, time.sleep(secs) only stalls the thread not the process according to the documentation.

Emilv2 commented:
If you are using the lock the update process will have to wait for the sleep to be finished and the lock to be released.

According to [this](https://stackoverflow.com/questions/56729764/python-3-7-asyncio-sleep-and-time-sleep) there is a difference between the two. `async.sleep()` also seems to be allowed in the HA code.

janiversen commented:
you are right my comment was wrong (and I deleted it), there is a difference between async and asyncio :-( 

I am testing async.sleep() right now.

janiversen commented:
Changed time.sleep to asyncio.sleep(), and at the same time learned something :-) it seems time.sleep() blocks the process in some versions of python and only the thread in other versions, at least that how I understand the documentation. @Emilv2 thanks for the suggestion. 

@bdraco I hope this solves your request for change, I made it as an extra commit to make the review easier, if you want me to squash it or make other changes please let me know.

Emilv2 commented:
If I understand it correctly `time.sleep()` does indeed only block the current thread but HA uses only one thread for updating components for simplicity (avoid the need of using locks etc) and relies on asyncio to speed up the task.

janiversen commented:
I will make that change, but I do not understand why it is needed? The code with a sync client works, it runs now on my production system as well as on my dev. system.

If I understand the doc correctly, you cannot have async call within a sync function but sync call within a async function does not seem to be deemed illegal.

bdraco commented:
> I will make that change, but I do not understand why it is needed? The code with a sync client works, it runs now on my production system as well as on my dev. system.
> 
> If I understand the doc correctly, you cannot have async call within a sync function but sync call within a async function does not seem to be deemed illegal.

If you do IO in the event loop it will block so everything has to wait until the IO is completed. 

https://docs.python.org/3/library/asyncio-eventloop.html

janiversen commented:
That  is a very good point, that I should have thought of. 

I tried to implement async modbus, this is however a bigger PR, because the current setup() is not async.

In order to make a balance between not changing everything and solve your valid change request, I modified the last commit to ensure all functions except client.connect() are synchronous like before this patch, and client.connect (which contains the asyncio.slepp()) are contained in the call:
asyncio.run_coroutine_threadsafe(client.connect(), hass.loop)
(ref: https://developers.home-assistant.io/docs/asyncio_working_with_async/#calling-async-functions-from-threads).

I hope this solves the problem, not in the most elegant way, but hopefully in an acceptable way.

Based on your input I have started to work on another patch (somewhat bigger, touching a lot of lines) that converts the whole setup to async, which I would like to submit later, together will solutions to some of the open issues. I am however not too sure if such a PR will be seen at "stepping on the toes of the CODEOWNER" ?

Emilv2 commented:
[According to the documentation](https://developers.home-assistant.io/docs/asyncio_working_with_async/) using `run_coroutine_threadsafe` will block the execution until it has finished, this is exactly what we want to avoid by using asyncio. The whole component should be changed to be async and use pymodbus async modbus client.

janiversen commented:
Yes it will block just like all the other calls are blocking, but sometimes one need to take smaller steps :-)

The current modbus implementation is synchronous and e.g. the connect() can take several seconds to complete, not to mention each read(), so we already have potential longer waits blocking the thread.
 
"just" using the async Modbus TCP client caused all kind of asserts, basically because setup() is synchronous, so setup() needs to be changed to async_setup() as well.

It seems you, bdraco and I are the only ones discussing, @adamchengtkc (CODEOWNER with a noreply mail address) seems to be inactive. Changing to async is a change that touches more or less all of the code, and in order to react swiftly to problems it should be done by the CODEOWNER, I am not sure how changing CODEOWNER is handled in this project.

My preference is to get this PR merged, and then make the async part as a separate bigger PR, followed by a couple of smaller PR to close the open issues. However if demanded I can also do it the other way round first make everything async and then solve the Huawei sun2000 problem.

bdraco commented:
@janiversen Blocking the event loop will affect all other integrations and home assistant from functioning normally.

Ideally, this would all get converted to async so it is much easier to address these types of issues.  That being said, I realize that is a big ask.

If you don't want to do that now, consider exploring the `select.select` on the socket to see if it is ready to write after the `connect` that I mentioned earlier instead.

janiversen commented:
@bdraco I am very much aware of that, which is the reason I was surprised to see a sync connect() with the default timeout.

select does not solve the problem at hand because the socket is ready for writing, but the app in the device needs time to initialize (in reality it is an error in sun2000).

I am a retired programmer with a 40+ years experience so “big ask” is a nice challenge. I will rewrite the code, give me a couple of days.

Emilv2 commented:
I can confirm that using `select.select` doesn't work. I look forward to see how the pymodbus component can be rewritten to async, I didn't manage to make it work with HA and I have no idea what went wrong.

janiversen committed:
modbus: bumb pymodbus version to 2.3.0

pymodbus version 1.5.2 did not support asyncio, and in general
the async handling have been improved a lot in version 2.3.0.

updated core/requirement*txt

janiversen committed:
modbus: change core connection to async

change setup() --> async_setup and update() --> async_update()

Use async_setup_platform() to complete the async connection to core.

listen for EVENT_HOMEASSISTANT_START happens in async_setup()
so it needs to be async_listen.

But listen for EVENT_HOMEASSISTANT_STOP happens in start_modbus()
which is a sync. function so it continues to be listen().

janiversen committed:
modbus: move setup of pymodbus into modbushub

setup of pymodbus is logically connected to the class modbushub,
therefore move it into the class.

Delay construction of pymodbus client until event
EVENT_HOMEASSISTANT_START arrives.

janiversen committed:
modbus: use pymodbus async library

convert pymodbus calls to refer to the async library.

Remark: connect() is no longer needed, it is done when constructing
the client. There are also automatic reconnect.

janiversen committed:
modbus: use async update for read/write

Use async functions for read/write from pymodbus.

change thread.Lock() to asyncio.Lock()

janiversen committed:
Modbus: patch for slow tcp equipment

When connecting, via Modbus-TCP, so some equipment (like the
huawei sun2000 inverter), they need time to prepare the protocol.

Solution is to add a asyncio.sleep(x) after the connect() and before
sending the first message.

Add optional parameter "delay" to Modbus configuration.
Default is 0, which means do not execute asyncio.sleep().

janiversen commented:
Finally !!! a highly changed integration, that are now fully async.

It turned out to be less simple as I assumed (devil stick in the details), good thing is that I had a chance of debugging lot of core functions, giving me a little bit of experience in how the core works.

This PR needs a good review, and I am afraid the current codeowner is not seeing this PR.

@Emilv2 can you please have a go with this PR on your system. You will see read errors from time to time (for me about once every hour), according to Huawei development (they are back working after a long vacation due to the virus), this happens if we read in parallel with a read from the management system, they will add some blocking in a later firmware.

@bdraco can you please take a look at the code, from the HA perspective. If I can ask a favor please also tell me, if there are inefficient/bad/... code.

Looking forward to see progress on this PR.

janiversen committed:
updated core/CODEOWNERS

committing result of 'python3 -m script.hassfest'.

janiversen commented:
how strange, I updated manifest.json but now CI tells me to also update core/CODEOWNERS and core/requirements*txt.

That really sounds like something that should be automated, in order to avoid conflicts between the PRs

bdraco commented:
> how strange, I updated manifest.json but now CI tells me to also update core/CODEOWNERS and core/requirements*txt.
> 
> That really sounds like something that should be automated, in order to avoid conflicts between the PRs

Use
`python3 -m script.hassfest`

janiversen commented:
and "python3 -m script.gen_requirements_all" already did, I hope CI passes now.

Thanks for the hint, I had to look at the raw CI log file to find the problems :-)

janiversen commented:
Help, pylint does not like the code:
homeassistant/components/modbus/__init__.py:208:12: E0633: Attempting to unpack a non-sequence defined at line 13 of pymodbus.client.asynchronous.tcp (unpacking-non-sequence)
homeassistant/components/modbus/__init__.py:217:12: E0633: Attempting to unpack a non-sequence defined at line 13 of pymodbus.client.asynchronous.tcp (unpacking-non-sequence)
homeassistant/components/modbus/__init__.py:225:12: E0633: Attempting to unpack a non-sequence defined at line 13 of pymodbus.client.asynchronous.udp (unpacking-non-sequence)

loop, client = fn() is the standard way in all examples, and furthermore this seems to be a pylint problem. See issue #2855
https://github.com/PyCQA/pylint/issues/2855

please advice, what I should do.

bdraco commented:
If pylint is doing the wrong thing you can disable it with a comment

`# pylint disable: https://github.com/PyCQA/pylint/issues/2855`

bdraco commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -178,73 +146,136 @@ def write_coil(service):
+    def __init__(self, client_config, main_loop):
         """Initialize the Modbus hub."""
-        self._client = modbus_client
-        self._lock = threading.Lock()
```
It looks like the lock was removed.  I think it needs to be converted to a `asyncio.Lock()` instead.

janiversen replied:
Done (will be uploaded when all changes are ready). I removed thread.lock() because I had a deadlock, and simply forgot to add async.Lock().

bdraco commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -178,73 +146,136 @@ def write_coil(service):
         client_name = service.data[ATTR_HUB]
         hub_collect[client_name].write_coil(unit, address, state)
```
All the writes are going to have to be async as well.

```
https://pymodbus.readthedocs.io/en/latest/source/example/async_asyncio_client.html
```

bdraco replied:
write_coil is going to need the same

janiversen replied:
have made all write async. I still have to test if I need to change hass.services.register call

bdraco replied:
If you are calling it from async use
`hass.services.async_register`

janiversen replied:
Thanks a lot, that call should be added to the async 101 page. I googled but did not find it.

bdraco commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -178,73 +146,136 @@ def write_coil(service):
+            await asyncio.sleep(self._config_delay)
+            self._config_delay = 0
+
+    def setup(self):
```
It doesn't look like there is any I/O here so you can mark this async safe with the @callback decorator
```
from homeassistant.core import callback
```
```suggestion
    @callback
    def async_setup(self):
```

janiversen replied:
There are actually IO involved the client* connect to the target. However for some (to me) unknown reason the client* cannot be called from an async method.

bdraco commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -178,73 +146,136 @@ def write_coil(service):
+        kwargs = {"unit": unit} if unit else {}
+        return await self._client.read_holding_registers(address, count, **kwargs)
 
     def write_coil(self, unit, address, value):
```
The writes will need to be converted to async as well like `def write_coil(service):`

janiversen replied:
done

bdraco commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -178,73 +146,136 @@ def write_coil(service):
+        kwargs = {"unit": unit} if unit else {}
+        self._client.write_coil(address, value, **kwargs)
 
     def write_register(self, unit, address, value):
```
The writes will need to be converted to async as well like `def write_coil(service):`

janiversen replied:
done

bdraco commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -178,73 +146,136 @@ def write_coil(service):
+        kwargs = {"unit": unit} if unit else {}
+        self._client.write_register(address, value, **kwargs)
 
     def write_registers(self, unit, address, values):
```
The writes will need to be converted to async as well like `def write_coil(service):`

janiversen replied:
done

bdraco reviewed (changes requested):
This is getting closer.  Good job!

bdraco commented:
@adamchengtkc If you have some spare cycles, it would really great if you could chime in as this is moving closer to merge each day

Emilv2 commented:
Working good for me on the Huawei  sun2000L. You can use ` _,  client = ` since `loop` is never needed (I don't think it's useful to add in the debug log since we're just using `hass.loop`).

adamchengtkc commented:
Sorry I am a bit busy at the moment. May be @vzahradnik can have a look? Please also consider adding yourself to the codeowner

vzahradnik commented:
> Sorry I am a bit busy at the moment. May be @vzahradnik can have a look? Please also consider adding yourself to the codeowner

@adamchengtkc I have no problem adding myself there. As a matter of fact, I planned to upstream several new Modbus entities, I just had no time to clean up the code.

I will look at this PR tomorrow.

janiversen commented:
@Emilv2 thanks for the test, I have changed to _. The reason I did not do it earlier, is I followed a bigger discussion whether using temp variables or _ was more correct, and it turns out that _ is actually a variable in python3.

@bdraco thanks for your review, I think I have resolved all your requests. I am still a bit unsure about hass.services.register but it seems to work with an async function.

 @vzahradnik looking forward to see your review. A lot of the code is changed with the update to async (reguested by @bdraco), my original change was a simple delay after connect.

Lets see if CI likes the new version.

janiversen committed:
Modbus: silence pylint false positive

pylint does not accept that a class construction __new__
can return a tuple.

vzahradnik commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -142,109 +109,179 @@ def stop_modbus(event):
-        """Connect client."""
-        with self._lock:
-            self._client.connect()
+        # with self._lock:
```
Please remove commented code.

vzahradnik commented:
@janiversen great job! Reworking the code to be async, and updating the `pymodbus` to the latest version was on my TODO list; you saved me quite some time :)

I gave you several comments in my review but all of them are related to the same thing - error handling. Please go through all Modbus entities and make sure all the calls are async. Several entities have local wrappers like `_write_register` or `_read_register` which were not reworked to async in all places.

Also, note that bulk of the code in these wrappers is in a `try...raise` block. Synchronous `Pymodbus` client raised a `ConnectionException` but other kinds of exceptions were returned as a normal result. Therefore, I introduced the `isinstance(result, (ModbusException, ExceptionResponse))` check below the `try...catch` block. It's used in several places; you can easily find all the occurrences.

Now, when you use an async client we need to find out how this error handling is affected, and adapt the code if necessary.

janiversen commented:
Thanks for the nice words. When I look at your review I can only see one comment (about commented out code) ?

in Async reconnect is automatically, but I think connectExcepcion is still issued.

I am aware of the wrapper functions, and I will do an extra walk through to secure I got them all converted.

I am currently having a small fight with the test, which does not work on GitHub. Tox is giving me problems when I try to run
tox -e py38 tests/components/modbus/test_sensor.py
I suppose I am doing something wrong.

vzahradnik reviewed (changes requested):
Please look at my comments. Thanks!

vzahradnik commented:
@janiversen you should be able to see my comments now. It was my first time creating a review and I didn't click on Submit.

As for errors, it might be that everything will work without a change but we need to test it anyway. When I worked on error handling, I had a local Modbus server which I interrupted to simulate various errors. My plan was to write unit tests one day but I improve this component one step at a time.

If you want, I can help with testing during the next week. At least you could rework all methods to be async (see my comments), then I can grab your code.

Regarding tox, I can't help much. Sometimes I too have problems with it. Maybe you could try to run the tests in isolated Docker container, as suggested in the docs:
https://developers.home-assistant.io/docs/development_environment

janiversen commented:
@vzahradnik Thanks a lot, I can see your comments now. I will rework the code, all read/write will be async, but some of the setup calls still needs to be sync.

I guess I have to rework the test code as well, I am trying to get my head around that. I will try the docker method. It turns out my tox problem is something basic, it tries to install all requirements and fails (how do you call tox?).

Would be nice if you can help test, this is a lot of changes so more test is better :-)

vzahradnik commented:
@janiversen usually I commit my changes and let Travis do the testing. I receive email after completion. When I tried to run tox locally, I dealt with issues. The Docker way was the most stable for me.

OK, please go through the comments. I'm sure the code I pointed out can be made async. Then I can help with testing.

Emilv2 commented:
I tested a bit more, and it fails on the Huawei sun2000L when the `scan_interval` is a bit higher than 10 (e.g. 60). `client.close()` followed by another sleep and `client.connect()` should be able to solve this I think. Or adding a keepalive as you mentioned before?

janiversen commented:
bigger scan intervals will lead to a disconnect unless keepalive is active. I have added this to my todo list. At least the “house arrest” here in Spain, means I can use more time on this PR.

janiversen committed:
Modbus: move constants to const.py

Create const.py with constants only used in
the modbus integration.

Duplicate entries are removed, but NOT any entry that would
lead to a configuration change.

Some entries were the same but with different names, in this
case renaming is done.

Also correct the tests.

janiversen committed:
Modbus: move connection error handling to ModbusHub

Connection error handling depends on the hub, not the
entity, therefore it is logical to have the handling in
ModbusHub.

All pymodbus call are added to 2 generic functions (read/write)
in order not to duplicate the error handling code.

Added property "available" to signal if the hub is connected.

janiversen commented:
@vzahradnik I have converted all to async (unless I have forgotten something), added a new const.py and made the error handling a lot more robust.

@Emilv2 this version should also solve your disconnect problem.

Looking forward to hear comments.

vzahradnik commented:
@janiversen I will grab your latest code and look through it. Also, I will verify if error handling still works properly.

Probably I'll take a look today or tomorrow. Then, I'll let you know in this PR.

Thanks!

janiversen commented:
Sounds good, take your time.

I tested using a local server, and tried different scenarios
- no server at startup
- server  goes away for a blip
- server goes away for a long time
etc.
But maybe you find other issues, at the end of the day it is important that we have a stable/robust integration.

There are a couple of CI issues, which I will solve today. I have also been thinking of moving the _read_register methods into modbus hub, to avoid duplicating code in all the entities.

janiversen committed:
Modbus: CI cleanup

Solve CI problems.

Emilv2 commented:
The disconnect problem with higher `scan_interval` problem is solved. How did you solve this? I can't find it in your code.

vzahradnik commented:
@janiversen hi, I briefly checked your code. I have few observations:
- Error handling: original behavior was to log error only once to not spam the log. Now, in __init__ you log each failed attempt. To log only once is one of the requirements to get Silver certification for the component: https://developers.home-assistant.io/docs/integration_quality_scale_index
- `consts.py`:
`REGISTER_TYPE_COILS` is not a register type, but a config array name. Also, it's deprecated. I would consider renaming it to `DEPRECATED_CONF_COILS` and moving into "deprecated" group
`REGISTER_TYPE_COIL` and `REGISTER_TYPE_DISCRETE`: In Modbus terminology, coils and discrete inputs are not registers. They really are "coils" and "discrete_inputs". I would rename these constants to make it clear.
- `CONF_DATA_TYPE` is defined twice in this file

Also, I've got several errors in the log file. Here they are:
```bash
2020-03-24 16:12:10 WARNING (MainThread) [pymodbus.client.asynchronous] Not Importing deprecated clients. Dependency Twisted is not Installed

/usr/lib64/python3.8/concurrent/futures/thread.py:82: RuntimeWarning: coroutine 'ModbusCoilSwitch.update' was never awaited
  del work_item
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
/usr/lib64/python3.8/asyncio/base_events.py:1844: RuntimeWarning: coroutine 'ModbusCoilSwitch.update' was never awaited
  handle = self._ready.popleft()
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
/usr/lib64/python3.8/asyncio/base_events.py:1860: RuntimeWarning: coroutine 'ModbusCoilSwitch.update' was never awaited
  handle = None  # Needed to break cycles when an exception occurs.
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

Please take a look at these issues. I can review your changes again tomorrow.

Thanks!

janiversen commented:
@vzahradnik thanks for taking the time. I will do the rename the REGISTER_* was to see which Pymodbus call should be used, and you are right the naming should be different.

I found out that there are at least 2 integrations that depend on modbus, so I needed to leave a couple of constants in __init__.py, that is going to be changed in a follow up PR.

How do you produce the messages ? can you e.g. let me see the part configuration.yaml (if that is how you got them) ?

The first warning is from Pymodbus, and according to a google search, you cannot get rid of it.

I am having a big fight with test_modbus_sensor.py, it breaks in line 79, due to an error in line 230 of sensor.py. It tells me readResult cannot be used (see the CI output). If you have a clue, please share it.

@Emilv2 Pymodbus v2.30 with async have keep alive as default (actually I do not know how to turn it off).

vzahradnik commented:
@janiversen my configuration is rather huge. I will try to strip it down and provide to you.

Regarding Pymodbus, does this mean we don't need Twisted library in your implementation? I'm not yet familiar with async Pymodbus client. If that's true, we should fix that on library level. For now, it would be great to at least create an issue for tracking purposes.

janiversen commented:
That would be real super (stripping the configuration). I have not installed twister, but I will check to see. According to the Pymodbus documentation, you can choose between asyncio, tornado and twister.

Once I get the other issues out of the way, I will file a issue with Pymodbus (which they probably then will ask me to fix :-) ).

REGISTER_TYPE_COILS is deprecated in binary_senser.py but used actively in switch.py, so
renaming to DEPRECATED_CONF_COILS, would not be correct (I think).

vzahradnik commented:
Oh, you're right! Because you united all the common constants, it is deprecated only in `binary_sensor`. Discrete inputs allow only reading of values, and therefore this input type is not implemented in `switch`, which needs to write the state, too.

janiversen commented:
It is nice to be right from time to time. I am renaming the others as I write this.

I united all constants, because it (me at least) is insecure to have the same constant defined multiple times, with a potential different value, and apart from that it can give a lot of confusion.

Any ideas in respect to me test_modbus_sensor.py problem ? ("await" and "readResult" are not compatible). There should be a new CI test log in a couple of minutes.

vzahradnik commented:
> 
> 
> It is nice to be right from time to time. I am renaming the others as I write this.
> 
> I united all constants, because it (me at least) is insecure to have the same constant defined multiple times, with a potential different value, and apart from that it can give a lot of confusion.
> 
> Any ideas in respect to me test_modbus_sensor.py problem ? ("await" and "readResult" are not compatible). There should be a new CI test log in a couple of minutes.

I'm not an expert to unit testing, Python is not even my primary language. However, I might take a look but first I want to strip my config and provide the details.

vzahradnik commented:
@janiversen hi, I stripped down my config to just a bare minimum:
https://gist.github.com/vzahradnik/fd913da6f27cfbaaf4296317cc961d24

Now I get three types of errors:
1. Platform error - probably a typo in the code
```bash
2020-03-25 21:08:10 ERROR (MainThread) [homeassistant.config] Platform error: binary_sensor
Traceback (most recent call last):
  File "/srv/git/hassio/home-assistant/homeassistant/config.py", line 769, in async_process_component_config
    platform = p_integration.get_platform(domain)
  File "/srv/git/hassio/home-assistant/homeassistant/loader.py", line 276, in get_platform
    cache[full_name] = importlib.import_module(
  File "/srv/git/hassio/home-assistant/venv/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/srv/git/hassio/home-assistant/homeassistant/components/modbus/binary_sensor.py", line 17, in <module>
    from .consts import (
ModuleNotFoundError: No module named 'homeassistant.components.modbus.consts'
```

2. Assertion error
```bash
2020-03-25 21:08:10 ERROR (MainThread) [homeassistant.components.sensor] Error while setting up modbus platform for sensor
Traceback (most recent call last):
  File "/srv/git/hassio/home-assistant/homeassistant/helpers/entity_platform.py", line 177, in _async_setup_platform
    await asyncio.wait_for(asyncio.shield(task), SLOW_SETUP_MAX_WAIT)
  File "/usr/lib64/python3.8/asyncio/tasks.py", line 483, in wait_for
    return fut.result()
  File "/srv/git/hassio/home-assistant/homeassistant/components/modbus/sensor.py", line 134, in async_setup_platform
    ModbusRegisterSensor(
  File "/srv/git/hassio/home-assistant/homeassistant/components/modbus/sensor.py", line 192, in __init__
    assert 1 == 0
AssertionError
```

3. Error with async handling in ModbusCoilSwitch:
```bash
/usr/lib64/python3.8/asyncio/base_events.py:1860: RuntimeWarning: coroutine 'ModbusCoilSwitch.update' was never awaited
  handle = None  # Needed to break cycles when an exception occurs.
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
/usr/lib64/python3.8/concurrent/futures/thread.py:82: RuntimeWarning: coroutine 'ModbusCoilSwitch.update' was never awaited
  del work_item
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
/usr/lib64/python3.8/asyncio/base_events.py:1844: RuntimeWarning: coroutine 'ModbusCoilSwitch.update' was never awaited
  handle = self._ready.popleft()
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

I hope you'll be able to simulate the same behavior.

janiversen commented:
@vzahradnik thanks  a lot, first error is solved (sorry about that, it was a typo).

I will look at second and third error tomorrow, and let you know when the code it ready for more testing.

janiversen commented:
@vzahradnik I am confused, I have been testing all day with your configuration.yaml and I get no errors ??

vzahradnik commented:
@janiversen it might be that this issue is affected also by the server. I use `pymodbus` running on local port. What do you use as an emulator?

I could provide my server testing script, or I can try debugging your code by myself. I should hopefully have some time during weekend.

janiversen commented:
I use "Pymodbus Asyncio Server Example" running on the local machine.

If you provide your script I will try with that instead. 

And if I cannot solve it or reproduce it, it would be nice if you can have a look in the weekend.

janiversen commented:
@vzahradnik you can mail it directly to me (might be easier) jancasacondor@gmail.com

vzahradnik commented:
> jancasacondor@gmail.com

Will do!

janiversen commented:
@vzahradnik I get the ModbusCoilSwitch.update, so let me debug to see why it works with my server. Will try to get it done tonight. Thanks for the help.

vzahradnik commented:
Great news!

Sent with [ProtonMail](https://protonmail.com) Secure Email.

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
On čtvrtek 26. březen 2020 19:01, jan iversen <notifications@github.com> wrote:

> [@vzahradnik](https://github.com/vzahradnik) I get the ModbusCoilSwitch.update, so let me debug to see why it works with my server. Will try to get it done tonight. Thanks for the help.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, [view it on GitHub](https://github.com/home-assistant/core/pull/32557#issuecomment-604585962), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/ABCBBXUGU6UI5JO3EKY5TU3RJOKBDANCNFSM4LDQWPZQ).

janiversen committed:
Modbus: remove close of client

close() no longer exist in the pymodbus library, use
del client instead.

janiversen commented:
the assert error in sensor.py seems to be in a previous version, at least I cannot find the assert.

found and corrected the problem in switch.py (update --> async_update).

@vzahradnik   Lets see what other errors you can come up with :-)

vzahradnik commented on homeassistant/components/modbus/const.py:
```diff
@@ -0,0 +1,72 @@
+
+# register types
+CALL_TYPE_COIL = "coil"
+CALL_TYPE_COILS = "coils"
```
This is actually an array, not an attribute. It's equivalent to the `CONF_INPUTS` field below. Consider renaming to `CONF_COILS`.

Here's an actual example:
```yaml
switch:
  platform: modbus
  coils:
    - name: plc_01_door_manual
      hub: plc_01
      slave: 1
      coil: 1
...
```

vzahradnik reviewed (changes requested):
Please take a look at my comments. Thanks!

vzahradnik commented:
@janiversen I tested and reviewed your latest code. Everything works for me now. I found a few cosmetic issues. Otherwise, I think this code is in a good shape, and from my point of view I will hapilly see this PR merged.

For a reference, I'm putting here log warning, which needs to be fixed in the `pymodbus` library someday:
```bash
/srv/git/hassio/home-assistant/venv/lib/python3.8/site-packages/pymodbus/client/asynchronous/asyncio/__init__.py:250: DeprecationWarning: "@coroutine" decorator is deprecated since Python 3.8, use "async def" instead
  def start(self, host, port=502):
```

janiversen commented:
I will make your suggested changes later. I have seen the warnings, but it is really a warning about it will not work in 3.10 so I do not think the pymodbus people will see it as sn urgent issue.

Thanks for your time so far, once all is cleaned up (tomorrow) I will ask to get merged.

janiversen committed:
Modbus: correct review comments

Adjust code based on review comments.

janiversen commented:
@bdraco finally this PR is robust and ready to merge (pending if you have new comments).

@vzahradnik have helped a lot with review (as pr. suggestion from  the code owner @adamchengtkc) and testing. Now we have an integration a lot more robust and totally async.

Looking forward to get this merged, so both @vzahradnik and myself can continue on follow-up PRs.

bdraco commented on homeassistant/components/modbus/binary_sensor.py:
```diff
@@ -109,33 +107,18 @@ def available(self) -> bool:
         self._value = result.bits[0]
         self._available = True
-
-    def _set_unavailable(self):
```
Ideally this retain the logging when it goes unavailable per https://developers.home-assistant.io/docs/integration_quality_scale_index/

"Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected."

janiversen replied:
Confusion, I think you are looking at an old push. _set_unavailable() no longer exist in the code. We should only log once for every problem.

bdraco commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -88,163 +97,219 @@
         """Stop Modbus service."""
         for client in hub_collect.values():
-            client.close()
+            del client
```
Is this going to shutdown cleanly? 

Does the underlying protocol need to be stopped here?

janiversen replied:
yes, there are no longer a .close() method. Pymodbus catches the del and does whatever needs to be done.

Need to, is a matter of taste, since hass is terminating (at least that is the only way I could provoke the stop event) everything will be shutdown automatically. However maybe we would want to be able to shutdown a single integration and then it is needed.

bdraco commented on homeassistant/components/modbus/climate.py:
```diff
@@ -270,25 +269,7 @@ def _read_register(self, register_type, register) -> Optional[float]:
+        await self._hub.write_registers(self._slave, register, [value, 0])
         self._available = True
-
-    def _set_unavailable(self, register):
```
Ideally this retain the logging (since its already in another place, it would be good to centralize this to keep the code DRY) when it goes unavailable per https://developers.home-assistant.io/docs/integration_quality_scale_index/

"Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected."

janiversen replied:
That should be done in the latest version (see my comment on git rebase, which might have caused you to look at not updated code).

bdraco commented on homeassistant/components/modbus/sensor.py:
```diff
@@ -252,16 +251,3 @@ def update(self):
 
         self._available = True
-
-    def _set_unavailable(self):
```
Please see comment in climate.py

janiversen replied:
Somehow there is a difference between my version (on my disk) and what you are looking at. Just to be sure, I will do another "git push -f". 

Wait a sec. can you be looking at something else, because you have loaded it as a branch. Be aware I use "git rebase" quite often (as long as a PR is open and not in final review) to keep the commits more logical.

bdraco commented on homeassistant/components/modbus/switch.py:
```diff
@@ -143,58 +146,38 @@ def available(self) -> bool:
+        await self._hub.write_coil(self._slave, coil, value)
         self._available = True
 
-    def _set_unavailable(self):
```
Please see comment in climate.py

janiversen replied:
same comment

bdraco commented on homeassistant/components/modbus/switch.py:
```diff
@@ -280,46 +263,28 @@ def update(self):
+        await self._hub.write_register(self._slave, self._register, value)
         self._available = True
-
-    def _set_unavailable(self):
```
Please see comment in climate.py

janiversen replied:
same comment.

bdraco commented on homeassistant/components/modbus/climate.py:
```diff
@@ -21,41 +21,42 @@
         vol.Required(CONF_NAME): cv.string,
         vol.Required(CONF_SLAVE): cv.positive_int,
         vol.Required(CONF_TARGET_TEMP): cv.positive_int,
-        vol.Optional(CONF_COUNT, default=2): cv.positive_int,
```
If we are renaming a config option, we need to label this as a breaking change so users know they need to update their configuration (or revert the name change)

janiversen replied:
We are renaming the constant, not the content of the constant, a slight difference. Seen from a use POW, there are no changes. (but a very good catch).

bdraco replied:
Got it 

Old
```
const.py:CONF_DATA_COUNT = "data_count"
```
New
```
climate.py:CONF_COUNT = "data_count"
```

bdraco reviewed (changes requested):
Hi @janiversen   Glad to see this is getting closer!

Would you please ensure all outstanding comments are resolved and mark them as resolved?

I did a quick run though.  I'll do a more through pass this weekend when I'm not focused on my day job.

janiversen commented:
Sorry if I made confusion by doing "git rebase", but it is natural to me, to update the commit and thus not first make something wrong and then correct it in a later commit. This way of working stops for me, when a PR is set for final review (as this one is now).

If "git rebase" causes too much confusion, please tell me, and I will refrain from using it.

bdraco commented:
> Sorry if I made confusion by doing "git rebase", but it is natural to me, to update the commit and thus not first make something wrong and then correct it in a later commit. This way of working stops for me, when a PR is set for final review (as this one is now).
> 
> If "git rebase" causes too much confusion, please tell me, and I will refrain from using it.

rebase should be fine.  I can't see where you are logging when the device goes unavailable, and logging where it is recovered.   I'm looking at 4b1c64febf590a296b53515899618180e6ca0fb2 as the sha when I check this out.  Can you point me to a specific file / line?

janiversen commented:
We are not doing that loging, because it is handled inside the Pymodbus library. It will automatically try to reconnect with a doubling delay for each failed attempt, so there are no spefic log for that.

Pymodbus will return an exception when it gives up, and that is handled in __init__.py line 255, which also handles spurious read errors.

Pymodbus will return None, if there are no connection, which will set unavailable at entity level.

bdraco commented on homeassistant/components/modbus/manifest.json:
```diff
@@ -2,7 +2,7 @@
   "name": "Modbus",
   "documentation": "https://www.home-assistant.io/integrations/modbus",
-  "requirements": ["pymodbus==1.5.2"],
+  "requirements": ["pymodbus==2.3.0", "twisted==20.3.0"],
```
It doesn't look like `twisted` is being used.

janiversen replied:
Correct, but if we do not install Pymodbus throws a warning about it missing, which I wanted to remove from the log. If you can accept the warning I will remove the requirement.

I have been mailing with the Pymodbus people, and they told me it is not so easy to solve as I thought, they are waiting for an upcoming python release.

Please advice, what you want me to do.

bdraco replied:
@balloob Seems like the warning is better than installing `twisted` but I'd prefer if you weighed in.

balloob replied:
Yeah totally, let's get the warning and then pymodbus should be fixed to not throw that warning.

janiversen replied:
@bdraco @balloob thanks for the input, I have removed the dependency on twister. Personally I was in doubt about the best road.

janiversen committed:
Modbus: remove twister dependency

Pymodbus in asyncio mode do not use twister but still throws a
warning if twister is not installed, this warning goes into
homeassistant.log and can thus cause confusion among users.

However installing twister just to avoid the warning is not
the best solution, therefore removing dependency on twister.

janiversen commented:
@bdraco do you have more review comments ?

bdraco commented:
I’m still planning on running though this today now that we hit the weekend and I can give it more attention

bdraco commented on tests/components/modbus/test_modbus_sensor.py:
```diff
@@ -89,9 +85,9 @@ def __init__(self, register_words):
-    entity_id = f"{SENSOR_DOMAIN}.{sensor_name}"
-    state = hass.states.get(entity_id).state
-    assert state == expected
+    # entity_id = f"{MODBUS_DOMAIN}.{sensor_name}"
```
Please remove the commented out code.

janiversen replied:
Done

janiversen committed:
Modbus: review, remove comments.

remove commented out code.

bdraco commented:
@janiversen LGTM. I don't see anything else besides the comments out code. Once that is fixed, I can approve.  Someone else will likely do another pass on this at some point before or after merge given the size.

janiversen commented:
@bdraco thanks a lot for your review. I am working on the followup PR, so I hope is not going to take ages :-)

vzahradnik commented:
Somebody should remove the flag "small-pr" :smile:

MartinHjelmare commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -88,163 +97,219 @@
-        self._client = modbus_client
-        self._lock = threading.Lock()
-        self._name = name
+        _LOGGER.debug("Preparing setup: %s", client_config)
```
We don't want side effects, eg logging, in init method. Please move it outside, or remove it.

MartinHjelmare commented on homeassistant/components/modbus/__init__.py:
```diff
@@ -88,163 +97,219 @@
+        async with self._lock:
+            kwargs = {"unit": unit} if unit else {}
+            result = await func(address, count, **kwargs)
+            if isinstance(result, (ModbusException, ExceptionResponse)):
```
It would be nice if the client func raised errors instead of returning them. I guess this is the library responsibility.

MartinHjelmare commented on homeassistant/components/modbus/binary_sensor.py:
```diff
@@ -56,7 +54,7 @@
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_platform(hass, config, add_entities, discovery_info=None):
```
Please rename `add_entities` to `async_add_entities`.

MartinHjelmare reviewed (commented):
Looks good! Just some small comments for a future PR.
